### ISSUE TITLE
Move type to "correct" file

### ISSUE THREAD

**Issue by sharwell** [DotNetAnalyzers/Proposals#13]
I propose a refactoring analyzer which is capable of handling two situations.
1. A single top-level type is defined in a file which does not match the name of the class
   
   This refactoring would rename the file to match the metadata name of the type it contains. For non-generic types, this is just the simple name of the type. For generic types, this is the simple name for the type followed by a <code>`</code> and then the generic arity.
2. Multiple top-level types are defined in a file
   
   For the type, if any, which matches the name of the file, no refactoring is proposed. For other types defined in the file, the refactoring would move the type to its own file based on the name of the type.

It would also be possible to create a diagnostic for these situations, which is applicable to projects where the situation should be proactively avoided.

**Comment by tugberkugurlu** [comment]
yes, yes, yes! [I want this](https://twitter.com/tourismgeek/status/532977341111087106)!

![enjoy](https://cloud.githubusercontent.com/assets/328122/5071384/f778ea96-6e78-11e4-903f-7039784a17b1.gif)

**Comment by tugberkugurlu** [comment]
A few questions and thoughts I have for this:
- Is it possible to pop up a UI during refactoring phase to make the user choose a location inside the solution?
- It also needs handle using statement refactorings if the file is moved under a different namespace.

**Comment by sharwell** [comment]
For _this_ refactoring, it would make sense to place the file in the same folder as the source file it was extracted from. A separate refactoring could be implemented for moving a file to a new folder based on the namespace of the type it contains.

**Comment by johannesegger** [comment]
I know this feature from ReSharper, I love it, and that's why I would like to take that one. Creating and transfering a repository seems to be [not that easy](https://github.com/DotNetAnalyzers/Proposals/issues/3#issuecomment-63242599), so I suggest you create one under DotNetAnalysers.

Should this be a diagnostic or a refactoring?
I personally would prefer a diagnostic as I think type name and file name should be identical, but maybe that is controversial?

I thought of naming it either _TypeToFile[Diagnostic|Refactoring]_, _TypeNameToFileName[Diagnostic|Refactoring]_, _ExtractTypeToFile[Diagnostic|Refactoring]_ or _ExtractType[Diagnostic|Refactoring]_, but I would hope that you or anyone else has some better ideas, because none of the above exactly matches your description.

**Comment by sharwell** [comment]
It's pretty easy to move; that was just the first time i tried to move someone else's repository. I wouldn't let that stop you, especially since I'm not available to create a new one here until next week.

**Comment by johannesegger** [comment]
Ahh ok, I've never done that before. It doesn't stop me, no worries :-), I always have my local repo.

**Comment by sharwell** [comment]
Cool. I look forward to seeing it! :+1:

**Comment by johannesegger** [comment]
Ok, so this project is currently hosted [here](https://github.com/eggapauli/ExtractTypeToFileDiagnostic) and of course everyone is invited to contribute.

**Comment by paulomorgado** [comment]
@eggapauli, I believe you forgot to remove the `MakeUppercaseAsync` method.

**Comment by johannesegger** [comment]
Oh you are right, thanks. I used this as a template, because I'm new to the Roslyn API.
